sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -983,7 +983,7 @@ void ProgramManager::flushSpecConstants(const program_impl &Prg,
     // caller hasn't provided the image object - find it
     { // make sure NativePrograms map access is synchronized
       ContextImplPtr Ctx = getSyclObjImpl(Prg.get_context());
-      auto LockGuard = Ctx->getKernelProgramCache().acquireCachedPrograms();
+      std::lock_guard<mutex_class> Lock(MMutex);
```

Comment:
AFAIK `flushSpecContants` function is called from within `build` function. The latter locks  `Ctx->getKernelProgramCache().acquireCachedPrograms()`. One should be pretty careful when entering a context with acquiring another mutex lock as it may lead to deadlock (see [this description](https://stackoverflow.com/a/42720663/1644482) for reference).

Reply:
I was thinking a lot about the possibility of deadlock and while I came to the conclusion that this is not possible here.

If we look at the places where `Ctx->getKernelProgramCache().acquireCachedPrograms()` is locked, there is 4 places:
1. before building `PiProgram`, when we try to emplace to `Cache` the future program build in `getOrBuild()`.

1. when one thread builds a `PiProgram` in `getOrBuild()` and internally calls the `flush_spec_constants()` function in `flushSpecConstants()`.

1. `set_spec_constant_impl()` from `program_impl` class.

1. `flush_spec_constants()` from `program_impl` class.

Even if in one of these cases the new mutex (`MMutex`) is locked by other thread, it will be free after writing to `NativePrograms` or reading from `NativePrograms`.
Notes: The new mutex (`MMutex`) will only be used to access `NativePrograms` and will return the resource as quickly as possible.


Do you see any reasons why there might be a dead-lock or just wanted me to make sure there isn't a dead-lock?

Reply:
> The new mutex (`MMutex`) will only be used to access `NativePrograms` and will return the resource as quickly as possible.

Should it be renamed to `MNativeProgramsMutex`, then?


> Do you see any reasons why there might be a dead-lock or just wanted me to make sure there isn't a dead-lock?

No, I don't see the reasons for deadlock but it's better to have a couple of opinions/conclusions  in such places.

Reply:
> Should it be renamed to MNativeProgramsMutex, then?

Ok, I will rename `MNativeProgramsMutex`.

> No, I don't see the reasons for deadlock but it's better to have a couple of opinions/conclusions in such places.

Sure, I completely agree.